pyomo/contrib/satsolver/satsolver.py

```diff
@@ -0,0 +1,273 @@
+        for v in self.variable_list:
+            string = string + v
+        string = string + "Bounds:\n"
+        for e in self.bounds_list:
+            string = string + e
+        string = string + "Expressions:\n"
+        for e in self.expression_list:
+            string = string + e
+        string = string + "Disjunctions:\n"
+        for djn in self.disjunctions_list:
+            string = string + "Disjunction: " + djn[0] + "\n"
+            for disj in djn[1]:
+                string = string + "  " + disj[0] + " : " + "\n"
+                for c in disj[1]:
+                    string = string + "    " + c + "\n"
+        return string
+
+    def __init__(self, model=None):
+        self.variable_label_map = SymbolMap(NumericLabeler('x'))
+        self.prefix_expr_list = self._get_default_functions()
+        self.variable_list = []
+        self.bounds_list = []
+        self.expression_list = []
+        self.disjunctions_list = []
+        self.walker = SMT_visitor(self.variable_label_map)
+        self.solver = z3.Solver()
+
+        if model is not None:
+            self._process_model(model)
+
+    # Set up functions to be added to beginning of string
+    def _get_default_functions(self):
+        default = list()
+        default.append("(define-fun exp ((x Real)) Real (^ 2.718281828459045 x))")
+        return default
+
+    # processes pyomo model into SMT model
+    def _process_model(self, model):
+        for v in model.component_data_objects(ctype=Var, descend_into=False):
+            smtstring = self.add_var(v)
+        for c in model.component_data_objects(ctype=Constraint, active=True):
+            self.add_expr(c.expr)
+        for djn in model.component_data_objects(ctype=Disjunction):
+            if djn.active:
+                self._process_active_disjunction(djn)
+            else:
+                self._process_inactive_disjunction(djn)
+
+    # define bound constraints
+    def _add_bound(self, var):
+        nm = self.variable_label_map.getSymbol(var)
+        lb = var.lb
+        ub = var.ub
+        if lb is not None:
+            self.bounds_list.append("(assert (>= " + nm + " " + str(lb) + "))\n")
+        if ub is not None:
+            self.bounds_list.append("(assert (<= " + nm + " " + str(ub) + "))\n")
+
+    # define variables
+    def add_var(self, var):
+        label = self.variable_label_map.getSymbol(var)
+        domain = type(var.domain)
+        if domain is RealSet:
+            self.variable_list.append("(declare-fun " + label + "() Real)\n")
+            self._add_bound(var)
+        elif domain is IntegerSet:
+            self.variable_list.append("(declare-fun " + label + "() Int)\n")
+            self._add_bound(var)
+        elif domain is BooleanSet:
+            self.variable_list.append("(declare-fun " + label + "() Int)\n")
+            self._add_bound(var)
+        else:
+            raise NotImplementedError("SMT cannot handle" + str(domain) + "variables")
```

Comment:
It is good that this is defensive, but there are a number of other domains that are commonly used that you will not catch.  In particular, NonNegativeReals and PositiveIntegers.  See pyomo/core/kernel/set_types.py

Reply:
I was under the impression that those would still be handled, since they are all instances of either of those 3 classes. I will add some tests that confirm this

Reply:
While it is true that things like NonNegativeReals are *instances* of RealSet, the test in your code is `is`, and `Reals is RealSet == False`.  I think what you want is:
```
if isinstance(domain, RealSet):
```
(and so on)

Reply:
Okay thanks! That makes sense

Reply:
@jsiirola What you say would be true if he had just done `var.domain is RealSet`. But for some reason, Sunjeev's logic works because of `domain = type(var.domain)`, which is a bit confusing offhand to me.

Reply:
Ahh - my mistake: I had missed that detail in his implementation.